mllib/src/main/scala/org/apache/spark/ml/recommendation/ALS.scala

```diff
@@ -88,23 +88,23 @@ private[recommendation] trait ALSParams extends ALSModelParams with HasMaxIter w
   def getRank: Int = $(rank)
 
   /**
-   * Param for number of user blocks (>= 1).
+   * Param for number of user blocks (>= 1, or -1 to auto-configure).
    * Default: 10
    * @group param
    */
   val numUserBlocks = new IntParam(this, "numUserBlocks", "number of user blocks",
-    ParamValidators.gtEq(1))
+    ParamValidators.or(ParamValidators.gtEq(1), ParamValidators.eq(-1)))
```

Comment:
Using the proposed `ParamValidators.or` and `ParamValidators.eq` to allow for a block size of >= 1 or -1 to auto-configure when training